plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryClient.java

```diff
@@ -51,30 +62,92 @@
+            if (remoteDatasetName == null) {
+                // this might have been just created, reload cache
+                Map<String, String> mapping = StreamSupport.stream(listDatasets(projectId).spliterator(), false)
+                        // will throw on collisions to avoid ambiguity
+                        .collect(toImmutableMap(dataset -> dataset.getDatasetId().getDataset().toLowerCase(ENGLISH), dataset -> dataset.getDatasetId().getDataset()));
+                remoteDatasetNames.putAll(mapping);
+            }
+            remoteDatasetName = remoteDatasetNames.getIfPresent(datasetName);
+            if (remoteDatasetName != null) {
+                return remoteDatasetName;
+            }
+            // TODO: what to throw if still not found?
+            throw new TrinoException(BIGQUERY_OBJECT_NOT_FOUND, "Failed to find remote dataset name: " + datasetName);
+        }
+        catch (RuntimeException e) {
+            throw new TrinoException(BIGQUERY_OBJECT_NOT_FOUND, "Failed to find remote dataset name: " + firstNonNull(e.getMessage(), e), e);
+        }
+    }
+
+    String toRemoteTableName(String projectId, String remoteDatasetName, String tableName)
+    {
+        requireNonNull(projectId, "projectId is null");
+        requireNonNull(remoteDatasetName, "remoteDatasetName is null");
+        requireNonNull(tableName, "tableName is null");
+        verify(CharMatcher.forPredicate(Character::isUpperCase).matchesNoneOf(tableName), "Expected table name from internal metadata to be lowercase: %s", tableName);
+        if (!caseInsensitiveNameMatching) {
+            return tableName;
+        }
+
+        try {
+            @Nullable Map<String, String> mapping = remoteTableNames.getIfPresent(remoteDatasetName);
+            if (mapping != null && !mapping.containsKey(tableName)) {
+                mapping = null;
+            }
+            if (mapping == null) {
+                // this might have been just created, reload cache
+                mapping = StreamSupport.stream(listTables(DatasetId.of(projectId, remoteDatasetName), TableDefinition.Type.TABLE, TableDefinition.Type.VIEW).spliterator(), false)
+                        // will throw on collisions to avoid ambiguity
+                        .collect(toImmutableMap(table -> table.getTableId().getTable().toLowerCase(ENGLISH), table -> table.getTableId().getTable()));
+                remoteTableNames.put(remoteDatasetName, mapping);
+            }
+            String remoteTableName = mapping.get(tableName);
+            if (remoteTableName != null) {
+                return remoteTableName;
+            }
+            // TODO: what to throw if still not found?
+            throw new TrinoException(BIGQUERY_OBJECT_NOT_FOUND, "Failed to find remote table name: " + tableName);
+        }
+        catch (RuntimeException e) {
+            throw new TrinoException(BIGQUERY_OBJECT_NOT_FOUND, "Failed to find remote table name: " + tableName);
```

Comment:
In my opinion, it would be better to include `e` as the 3rd args. Otherwise, it's little hard to understand the cause even using debug log when key collision happened in L132.